network/src/sync/master.rs

```diff
@@ -0,0 +1,304 @@
+    node: Node<S>,
+    incoming: mpsc::Receiver<SyncInbound>,
+}
+
+struct SyncBlock {
+    address: SocketAddr,
+    block: Vec<u8>,
+}
+
+impl<S: Storage + Send + Sync + 'static> SyncMaster<S> {
+    pub fn new(node: Node<S>) -> (Self, mpsc::Sender<SyncInbound>) {
+        let (sender, receiver) = mpsc::channel(256);
+        let new = Self {
+            node,
+            incoming: receiver,
+        };
+        (new, sender)
+    }
+
+    async fn find_sync_nodes(&mut self) -> Vec<Peer> {
+        let our_block_height = self.node.expect_sync().current_block_height();
+        let mut interesting_peers = vec![];
+        for node in self.node.peer_book.connected_peers_snapshot().await {
+            if !node.judge() && node.quality.block_height > our_block_height + 1 {
+                interesting_peers.push(node);
+            }
+        }
+        interesting_peers.sort_by(|x, y| y.quality.block_height.cmp(&x.quality.block_height));
+
+        // trim nodes close to us if any are > 10 blocks ahead
+        if let Some(i) = interesting_peers
+            .iter()
+            .position(|x| x.quality.block_height <= our_block_height + 10)
+        {
+            interesting_peers.truncate(i + 1);
+        }
+
+        interesting_peers
+    }
+
+    async fn block_locator_hashes(&mut self) -> Vec<BlockHeaderHash> {
+        match self.node.expect_sync().storage().get_block_locator_hashes() {
+            Ok(block_locator_hashes) => block_locator_hashes,
+            _ => {
+                error!("Unable to get block locator hashes from storage");
+                vec![]
+            }
+        }
+    }
+
+    async fn send_sync_messages(&mut self) -> usize {
+        let sync_nodes = self.find_sync_nodes().await;
+
+        info!("requested block information from {} peers", sync_nodes.len());
+        let block_locator_hashes = self.block_locator_hashes().await;
+        let mut sent = 0usize;
+        let mut future_set = vec![];
+        for peer in sync_nodes.iter() {
+            if let Some(handle) = self.node.peer_book.get_peer_handle(peer.address) {
+                let block_locator_hashes = block_locator_hashes.clone();
+                sent += 1;
+                future_set.push(async move {
+                    handle.send_payload(Payload::GetSync(block_locator_hashes)).await;
+                });
+            }
+        }
+        futures::future::join_all(future_set).await;
+        sent
+    }
+
+    async fn receive_messages<F: FnMut(SyncInbound) -> bool>(&mut self, timeout_sec: u64, mut handler: F) {
+        let end = Instant::now() + Duration::from_secs(timeout_sec);
+        loop {
+            let timeout = tokio::time::sleep_until(end).fuse();
+            pin_mut!(timeout);
+            select! {
+                msg = self.incoming.recv().fuse() => {
+                    if msg.is_none() {
+                        break;
+                    }
+                    if handler(msg.unwrap()) {
```

Comment:
this condition is somewhat unclear - it seems that `handler` returning `true` indicates an error - this should be documented, or the name of the closure adjusted so it's more apparent

Reply:
it's not an error, but a terminator. `true` to return. it's an internal method, so i didn't see a need for documenting this, but i'll add some for clarity.